while-loops/problem-set-while-loops.md

```diff
@@ -0,0 +1,849 @@
+- follows the chain of indices to the end, and
+- returns the position of the end of the chain
+
+You may assume that `number_list` will never be an empty list, and that all of the values in the list are valid indices in the list.
+
+##### !end-question
+
+##### !placeholder
+
+```python
+def sidewinder(number_list):
+    pass
+```
+
+##### !end-placeholder
+
+##### !tests
+
+```python
+import unittest
+from main import sidewinder
+
+class TestPython1(unittest.TestCase):
+  def test_one(self):
+    self.assertEqual(0, sidewinder([0]))
+
+  def test_two(self):
+    self.assertEqual(1, sidewinder([2, 0, 1]))
+
+  def test_three(self):
+    self.assertEqual(10, sidewinder([4, 9, 0, 0, 1, 8, 0, 0, 11, 5, 0, 10]))
+```
+
+##### !end-tests
+
+##### !hint
+
+It may be useful to print out the positions you visit and the final position where the end of the list is to help think about this problem. Any print statements you add will not affect the test results. Even if you _do_ print the final position, be sure to also _return_ it!
+
+##### !end-hint
+
+##### !explanation
+
+An example of a working implementation:
+
+```python
+def sidewinder(number_list):
+    # set our starting position
+    next_position = 0
+
+    # iterate as long as the next position value is not 0
+    while number_list[next_position] != 0:
+        # update the next position value
+        next_position = number_list[next_position]
+
+    # return the position of the end of the chain
+    return next_position
+```
+
+Read through this version and compare to yours. What's the same? What's different? Are there even more ways to approach this problem?
+
+<br />
+
+For extra challenge, consider how we might handle bad inputs, like a missing or empty list. What if the list _does_ contain values that aren't valid indices? Or what if there's a loop in the indices like in `[1, 2, 1]`. Can you write a version of `sidewinder` which instead of returning the end of the chain, returns `True` if it visits _every_ position in the list, and `False` otherwise?
+
+<br />
+
+Be creative!
+
+##### !end-explanation
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+<!-- prettier-ignore-end -->
+
+### A Little Helper for the Next Question
+
+The following coding question uses a helper function called `input_int`. It works just like the [`input`](https://docs.python.org/3/library/functions.html#input) function that we've seen before, except that instead of returning a string, we will assume that it returns an integer.
+
+In the question block below, we don't need to define `input_int`. We can assume that it is available, the same way `input` is available.
+
+To experiment with the coding question in VS Code or repl.it, we will need an implementation for `input_int`. In Python, safely converting from a string to an integer requires handling exceptions. We haven't discussed that material yet, so here is a working implementation.
```

Comment:
I'm suggesting some wording changes to help angle this moreso from "Here's stuff you need to know for the next question to," "The next question is special because we're providing this. You should use it for the next question," in an attempt to strongly state "what you need is to use it, not to know it"

Summary of suggestions:

- Changes header (and content) tone to emphasize that it's provided and it's for the next question
- Changes tone to "it's already there, this is an FYI"
- Surfaces the responsibility of this function/why the student should care about this function first
- Give context to why it matters that we shouldn't define it and why we can assume it's available-- it's because students should use it
- removes the assumption that 1. this lesson always comes before handling exceptions and 2. that students haven't independently learned about that anyway

```suggestion
### Provided Helper Function for the Next Question

In the following coding question, there is a helper function called `input_int`. This function is responsible for getting user input, and it will always return an integer. (It works similarly to the [`input`](https://docs.python.org/3/library/functions.html#input) function that we've seen before, except the function does not return a string.)

In the question block below, `input_int` is already defined and available to use in our code. Do not define `input_int`. Use it because it is available, the same way `input` is available.

If you would like to experiment with the coding question in VS Code or repl.it, you will need an implementation for `input_int`. Here is a working implementation for you to copy/paste and use in VS Code or repl.it. It may contain concepts not covered yet, such as handling exceptions.
```

Reply:
replit.com would future-proof us, yeah!!!!

I'm manually editing my suggestion above to that, too, in case that's helpful.

Reply:
tanx